core/trino-main/src/main/java/io/trino/operator/LeafTableFunctionOperator.java

```diff
@@ -170,25 +167,19 @@ public Page getOutput()
                 // no more splits to process at the moment
                 return null;
             }
-            currentSplit = pendingSplits.remove();
-            resetProcessor();
-        }
-        else {
-            // a split is being processed
-            requireNonNull(currentSplit, "currentSplit is null");
+            ConnectorSplit nextSplit = pendingSplits.remove();
+            resetProcessor(nextSplit);
         }
 
-        TableFunctionProcessorState state = processor.process(processorUsedData ? null : currentSplit);
+        TableFunctionProcessorState state = processor.process();
         if (state == FINISHED) {
             processorFinishedSplit = true;
         }
         if (state instanceof Blocked blocked) {
             processorBlocked = toListenableFuture(blocked.getFuture());
         }
         if (state instanceof Processed processed) {
-            if (processed.isUsedInput()) {
-                processorUsedData = true;
-            }
+            checkState(!processed.isUsedInput(), "Invalid state, as no input has been provided: %s", state);
             if (processed.getResult() != null) {
```

Comment:
After your change the result cannot be null. `if` is redundant, you can optionally add a rnn.

Reply:
the API of `Processed` doesn't make it obvious that `!usedInput => has page`. In fact, we may want to support yielding, so not producing a page should be fine. however, the `if` is redundant anyway. if we don't return here, we return null 2 lines later. will remove